zaza/openstack/charm_tests/heat/tests.py

```diff
@@ -0,0 +1,348 @@
+#!/usr/bin/env python3
+#
+# Copyright 2019 Canonical Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Encapsulate heat testing."""
+import logging
+import json
+import os
+import subprocess
+from urllib import parse as urlparse
+from heatclient.common import template_utils
+from novaclient import exceptions
+
+import zaza.model
+import zaza.openstack.charm_tests.test_utils as test_utils
+import zaza.openstack.utilities.openstack as openstack_utils
+import zaza.charm_lifecycle.utils as charm_lifecycle_utils
+
+# Resource and name constants
+IMAGE_NAME = 'cirros-image-1'
+KEYPAIR_NAME = 'testkey'
+STACK_NAME = 'hello_world'
+RESOURCE_TYPE = 'server'
+TEMPLATES_PATH = 'files'
+FLAVOR_NAME = 'm1.tiny'
+
+
+class HeatBasicDeployment(test_utils.OpenStackBaseTest):
+    """Encapsulate Heat tests."""
+
+    @classmethod
+    def setUpClass(cls):
+        """Run class setup for running Heat tests."""
+        super(HeatBasicDeployment, cls).setUpClass()
+        cls.application = 'heat'
+        cls.keystone_session = openstack_utils.get_overcloud_keystone_session()
+        cls.heat_client = openstack_utils.get_heat_session_client(
+            cls.keystone_session)
+        cls.glance_client = openstack_utils.get_glance_session_client(
+            cls.keystone_session)
+        cls.nova_client = openstack_utils.get_nova_session_client(
+            cls.keystone_session)
+
+    @property
+    def services(self):
+        """Return a list services for OpenStack release.
+
+        :returns: List of services
+        :rtype: [str]
+        """
+        services = ['heat-api', 'heat-api-cfn', 'heat-engine']
+        return services
+
+    def _image_create(self):
```

Comment:
It feels like you're re-creating something that already exists in zaza-openstack-tests:

https://github.com/openstack-charmers/zaza-openstack-tests/blob/0c2f36fb4e7313e78b6982a594bbf595bc263c8e/zaza/openstack/charm_tests/glance/setup.py#L65-L77

Reply:
@ryan-beisner I understand the desire to use existing libraries/helper routines, but I thought that's what `openstack_utils` was for. I assumed that every set of tests in `zaza/openstack/charm_tests` was independent of each other, but now heat will be dependent on the glance setup test routines. Anyway, I will do what is best long-term, I will reuse helper functions from glance and nova as suggested.

Reply:
One more thing, I am also concerned about the functional test failure in the proposed tests, but I am using the exact same `openstack_utils` functions as the glance helper routine is, so I don't see that changing with the glance helper routine.